packages/fixie/src/auth.ts

```diff
@@ -0,0 +1,207 @@
+  const useApiKey = process.env.FIXIE_API_KEY ?? config.apiKey;
+  if (!useApiKey) {
+    // No key available. Need to punt.
+    return null;
+  }
+  const client = FixieClient.Create(useApiUrl, useApiKey);
+  const userInfo = await client.userInfo();
+  if (userInfo.is_anonymous) {
+    return null;
+  }
+  return client;
+}
+
+/** Returns an authenticated FixieClient, starting an OAuth flow to authenticate the user if necessary. */
+export async function AuthenticateOrLogin({
+  apiUrl,
+  configFile,
+  forceReauth,
+}: {
+  apiUrl?: string;
+  configFile?: string;
+  forceReauth?: boolean;
+}): Promise<FixieClient> {
+  if (!forceReauth) {
+    const client = await Authenticate({
+      apiUrl,
+      configFile,
+    });
+    if (client) {
+      const userInfo = await client.userInfo();
+      if (!userInfo.is_anonymous) {
+        return client;
+      }
+    }
+  }
+
+  const apiKey = await oauthFlow(apiUrl ?? FIXIE_API_URL);
+  const config: FixieConfig = {
+    apiUrl: apiUrl ?? FIXIE_API_URL,
+    apiKey,
+  };
+  SaveConfig(config, configFile ?? FIXIE_CONFIG_FILE);
+  const client = await Authenticate({ apiUrl, configFile: configFile ?? FIXIE_CONFIG_FILE });
+  if (!client) {
+    throw new Error('Failed to authenticate');
+  }
+  const userInfo = await client.userInfo();
+  term('🎉 Successfully logged into ')
+    .green(apiUrl ?? FIXIE_API_URL)(' as ')
+    .green(userInfo.username)('\n');
+  return client;
+}
+
+// The Fixie CLI client ID.
+const CLIENT_ID = 'II4FM6ToxVwSKB6DW1r114AKAuSnuZEgYehEBB-5WQA';
+// The scopes requested by the OAUth flow.
+const SCOPES = ['api-access'];
+
+/**
+ * Runs an interactive authorization flow with the user, returning a Fixie API key
+ * if successful.
+ */
+async function oauthFlow(apiUrl: string): Promise<string> {
+  const port = await findFreePort();
+  const redirectUri = `http://localhost:${port}`;
+  const state = crypto.randomBytes(16).toString('base64url');
+  const url = `${apiUrl}/authorize?client_id=${CLIENT_ID}&scope=${SCOPES.join(
+    ' '
+  )}&state=${state}&redirect_uri=${encodeURIComponent(redirectUri)}&response_type=code`;
+
+  const serverPromise = new Promise((resolve, reject) => {
+    const server = http
+      .createServer(async (req, res) => {
+        if (req.url) {
+          const searchParams = new URL(req.url, `http://localhost:${port}`).searchParams;
+          const code = searchParams.get('code');
+          const receivedState = searchParams.get('state');
+          if (code && receivedState === state) {
+            try {
+              const bodyFormData = new FormData();
+              bodyFormData.append('code', code);
+              bodyFormData.append('redirect_uri', redirectUri);
+              bodyFormData.append('client_id', CLIENT_ID);
+              bodyFormData.append('grant_type', 'authorization_code');
+              const response = await axios.post(`${apiUrl}/access/token`, bodyFormData, {
```

Comment:
You can probably use native fetch and don't need axios but not worth changing now